### PR TITLE
[IMP] account: performance improvement

### PR DESCRIPTION
**Pull request by Gorash**
'_get_fpos_by_region' is called quite regularly (once per product to
compute the price on website "/shop" for example). This is master data
that hardly changes, the cache saves a significant amount of time.
Refactoring how to get information might improve performance and no
longer require this cache.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Gorash** [comment]
Cache invalidation are only on master data which normally never changes. And so there is no negative impact on performance

File: addons/account/models/partner.py [hunk 1]
```diff
@@ -113,9 +114,15 @@ def write(self, vals):
         if zip_from or zip_to:
             for rec in self:
                 vals['zip_from'], vals['zip_to'] = self._convert_zip_values(zip_from or rec.zip_from, zip_to or rec.zip_to)
+        self.clear_caches()
```
Comment (oco-odoo) [inline addons/account/models/partner.py:117]: It's probably better not to do that all the time. 
If we are not changing the region selection fields of the fiscal position (country, state, zip range) or auto_apply is  False, we don't need to clear the cache, right ?

File: addons/account/models/partner.py [hunk 2]
```diff
@@ -105,6 +105,7 @@ def create(self, vals):
         zip_to = vals.get('zip_to')
         if zip_from and zip_to:
             vals['zip_from'], vals['zip_to'] = self._convert_zip_values(zip_from, zip_to)
+        self.clear_caches()
```
Comment (oco-odoo) [inline addons/account/models/partner.py:108]: We should only do that if the new fiscal position has auto_apply==True.

**Review by oco-odoo — CHANGES_REQUESTED** [review]
Two small things, then it's ok to me.

File: addons/account/models/partner.py (hunk 1 above)
Comment (Gorash) [reply addons/account/models/partner.py:117]: Indeed, but from what I have been told, this data only changes once a year, and therefore it does not really have an impact. I can indeed filter for cleaning the cache.

File: addons/account/models/partner.py (hunk 2 above)
Comment (Gorash) [reply addons/account/models/partner.py:108]: of course, I'll change it.

File: addons/account/models/partner.py [hunk 3]
```diff
@@ -113,9 +114,15 @@ def write(self, vals):
         if zip_from or zip_to:
             for rec in self:
                 vals['zip_from'], vals['zip_to'] = self._convert_zip_values(zip_from or rec.zip_from, zip_to or rec.zip_to)
+        self.clear_caches()
         return super(AccountFiscalPosition, self).write(vals)
 
+    def unlink(self):
+        self.clear_caches()
+        return super(AccountFiscalPosition, self).unlink()
```
Comment (fmdl) [inline addons/account/models/partner.py:122]: `return super().unlink()`

File: addons/account/models/partner.py [hunk 4]
```diff
@@ -181,13 +190,13 @@ def get_fiscal_position(self, partner_id, delivery_id=None):
...
+        fp_id = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, vat_required)
 
         # Then if VAT required found no match, try positions that do not require it
-        if not fp and vat_required:
-            fp = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, False)
+        if not fp_id and vat_required:
+            fp_id = self._get_fpos_by_region(delivery.country_id.id, delivery.state_id.id, delivery.zip, False)
 
-        return fp or self.env['account.fiscal.position']
+        return fp_id and Fp.browse(fp_id) or Fp
```
Comment (fmdl) [inline addons/account/models/partner.py:199]: `return Fp.browse(fp_id)` (if bool(fp_id) == False browse return an empty recordset.

File: odoo/addons/base/models/res_country.py [hunk 5]
```diff
@@ -129,6 +135,19 @@ class CountryGroup(models.Model):
     country_ids = fields.Many2many('res.country', 'res_country_res_country_group_rel',
                                    'res_country_group_id', 'res_country_id', string='Countries')
 
+    @api.model
+    def create(self, vals):
+        self.clear_caches()
+        return super(CountryGroup, self).create(vals)
```
Comment (fmdl) [inline odoo/addons/base/models/res_country.py:141]: `return super().create(vals)`

File: addons/account/models/partner.py (hunk 1 above)
Comment (oco-odoo) [reply addons/account/models/partner.py:117]: It changes when someone reconfigures something. So, indeed, not the most common operation, but I guess we should gain what we can :)

Comment (Gorash) [reply addons/account/models/partner.py:117]: At the expense of readability for one invalidation per year? I looked around a bit, but it makes for a weird condition and if someone changes relationships or else they will have to change the condition (without necessarily seeing the error as it is only when when there is a specific change that must invalidate the cache)

File: addons/account/models/partner.py (hunk 2 above)
Comment (Gorash) [reply addons/account/models/partner.py:108]: If someone sets the auto_apply to default = True then this invalidation might be wrong and the developer won't see it. It's identical to the other side, I think it is better to favor robustness with a tiny loss of performance (compared to the gain of the current pr)

**Comment by Gorash** [comment]
fp proposes to test a pure sql instead and not to cache.

**Pull request closed without merging**